[DOC] Streamline release process, add checklist to RTD

emdupre opened the pull request:
<!---
This is a suggested pull request template for tedana.
It's designed to capture information we've found to be useful in reviewing pull requests.

If there is other information that would be helpful to include, please don't hesitate to add it!

Please also label your pull request with the relevant tags.
See here for more information and a list of available options:
http://tedana.readthedocs.io/en/latest/contributing.html#pull-requests
-->

Closes #98.

Changes proposed in this pull request:

- Adds a release checklist to the ReadTheDocs site
- Add `release-drafter` bot to this repository
- Adopt `versioneer` for updating version numbers
- Incorporate Chris's `zenodo.js` script for updating RTD site with correct zenodo citation

emdupre commented:
.@kirstiejane, WDYT of these guidelines ? These are the steps I've followed in the past, but happy to add more background / detail as you think is appropriate !

tsalo commented:
Can/should we use a .zenodo.json file to specify the authors (e.g., [like Nipype does](https://github.com/nipy/nipype/blob/d2abc85cd98ea3a168a53306700312ca50eea87d/.zenodo.json))? We'll need authorship criteria, which I know have been discussed but which I don't think were ever finalized.

Also, should we include a CHANGELOG in the repository or do you think that logging changes in the release notes is enough? I've seen CHANGELOGs rendered on documentation sites (and have done this myself on [rapidtide](http://rapidtide.readthedocs.io/en/latest/whats_new.html) and [pybids](https://incf.github.io/pybids/whats_new.html), but I don't know if it's necessarily useful information for people using the doc site.

We also discussed on #57 the possibility of using the version-specific DOI in the duecredit citation for the repository. Is that possible or must the version be minted before we can get a DOI?

emdupre commented:
I'm not a big fan of the .zenodo.json file, personally, just because I think authorship this early on is not an easy decision, and the .zenodo.json file suggests a level of transparency that (I think) doesn't reflect reality. Also, I've seen contributors on other projects hesitate to add themselves if they've made "too small" of a contribution, but that self-imposed threshold varies significantly across people so there is a certain amount of arbitrariness in who ends up listed there.

At this point we have so few versions that I think I'm fine either way (i.e., in the release notes or in a proper CHANGELOG). Do you have a preference ?

 My understanding of the version-specific DOI is that we will need to mint the version before getting the DOI (@KirstieJane should correct me if I'm wrong !). I did, though, forget to add the duecredit citation update as a step -- I'll fix that, here !

Thanks for the review !

tsalo commented:
I think we need _something_ (not necessarily the zenodo json) to manage authors, even if it's managed purely by the maintainers. That way we can **1)** remove bots/integrations (which I think the Gitter Badger is), **2)** add full names, ORCIDs, affiliations, etc. when we have that info (e.g., for TomMaullin, who only shows up as his username), and **3)** manually include contributors that don't show up by default (e.g., @KirstieJane).

I don't think the CHANGELOG is necessary. It might make it easier to compile the changes for each release, but updating it would add a fair amount of overhead to each PR and is really easy to forget. Ultimately I guess it depends on when you want to deal with making sense of changes.

If we can't have the citation for a given version match that version, then maybe we should use the concept citation. The way it's currently set up (release --> get DOI --> update __init__ with new DOI), each version would cite the immediately preceding version.

emdupre commented:
.@tsalo and I were able to touch base about this in person at #NI2018, so to summarize:
1) The zenodo author list was set in the first version, and it was my oversight when minting subsequent DOIs to make sure that these matched. I'm updating the `v0.0.2` and `v0.0.3` release author listings to match the first version -- we'll add other authors, currently at my discretion, though as we get bigger we'll move to a more standardized process for this. For now, I'll add updating the zenodo author listing to the release checklist guidelines. 

2) To address the changelog comment, we could use Oscar's [update_changes](https://github.com/poldracklab/fmriprep/blob/master/update_changes.sh) bash script. I'm still not yet convinced we need it, but if it's easy to implement I suppose it can't hurt !

3) @chrisfilo pointed out that he has a javascript file to update the sphinx build with the correct Zenodo DOI. Is there any chance you could point out where that is, Chris?

emdupre commented:
I looked into this more and changed my mind on including a zenodo.json. We just won't advertise that this can be updated (yet).

I also found [Versioneer](https://github.com/warner/python-versioneer) from fMRIPrep, which seems like a great way to do version updates (though admittedly not in the ReadTheDocs site). I'll try to work on incorporating that next, though I don't think its incorporation should hold up this PR !

LMK what you think @tsalo @KirstieJane (and of course anyone else as well !).

tsalo commented:
Can you incorporate ORCIDs into the author list as well?

Also, on a related note, but for the distant future, does anyone know of any good example repositories with explicit authorship criteria explained anywhere? It's obviously not something we have to deal with now, but I am curious just to know what that looks like.

I don't know much about versioner, but it's used in pybids and seemed to be pretty cool. :+1:

chrisfilo commented:
Zenodo.js: https://github.com/chrisfilo/zenodo.js
Example usage: https://github.com/poldracklab/fmriprep/blob/master/docs/citing.rst

emdupre commented:
Thanks so much, @chrisfilo ! I've incorporated it into our RTD build and it works beautifully.

@tsalo @KirstieJane, I've simplified the instructions since now we've hooked so many great tools (specifically `zenodo.js`, `update_changes.sh`, and `versioneer`, all adapted from the fMRIPrep workflow) to make this as easy as possible. If this passes CI and looks ok to you, I think it's ready to merge !

tsalo commented on docs/contributing.rst:
``````````````````diff
@@ -42,3 +42,27 @@ Pull requests should be submitted early and often!
+`````````````````
+
+This is the checklist of items that must be completed when cutting a new release of tedana.
+These steps can only be completed by a project maintainer, but they a good resource for
``````````````````
Small typo: "they a good resource" --> "they are a good resource"

tsalo commented on docs/contributing.rst:
```diff
@@ -42,3 +42,27 @@ Pull requests should be submitted early and often!
+These steps can only be completed by a project maintainer, but they a good resource for
+releasing your own Python projects!
+
+    #. All continuous integration must be passing and docs must be building successfully
```
Missing period at end.

tsalo commented on docs/contributing.rst:
```diff
@@ -42,3 +42,27 @@ Pull requests should be submitted early and often!
+
+We have set up tedana so that releases automatically mint a new DOI with Zenodo;
+a guide for doing this integration is available `here`_.
+You can also `learn more about DueCredit`_ and why citing software is important.
```
This kind of comes out of nowhere now, given that the duecredit citation for the package no longer needs to be updated manually.

emdupre commented:
Ok, merging this ! :tada: Thanks, all !!

KirstieJane commented:
This is *so wild*!! I'm so sorry I'm so behind on everything. There's so much going on here - I love it!

Could we add to the to do list some sort of blog post highlighting all the different components? In particular I have no idea what the release drafter bot is doing but I like it's initiative :wink:

Not high on the priority list - and way to go on making this so slick! (To the tedada:tada: team and to @chrisfilo & @oesteban as they started it!!)
